### DISCUSSION TITLE
git_annotated_commit.parents type

### DISCUSSION THREAD

**Discussion by punkymaniac** [libgit2/libgit2#6048] [Q&A] [answered]
HI everyone,

Why is `git_annotated_commit.parents` not a `git_oidarray` type ?

In another way, `git_oidarray` and `git_array_oid_t` are basically the same kind of struct, so which one is better to use ? why keep both ?

**Comment by ethomson** [comment, marked as answer]
Great question!  `git_array_oid_t` is an internal type - we have a [`git_array_t`](https://github.com/libgit2/libgit2/blob/main/src/array.h) that we use as a base type for internal arrays, which allows us to perform common tasks (`push`, `pop`, etc) on an array represented as a contiguous chunk of memory for its members (as opposed to a `git_vector_t` which is an array of pointers).

So this is an internal utility class.  We _could_ expose this, but would prefer not to.  We don't want to give people general utility type functionality.  They should never want to `git_array_grow`, for example.  And making this a piecemeal situation where we expose _some_ of the `git_array_t` but not others is also a problem, where scope creeps over time and eventually, well, we're exposing the whole utility type functionality.

`git_oidarray` is strictly a return type.  It gives the users an array of oids, their size, and a single function to free them.  It's much easier to keep these two separate - even though they're somewhat similar - than to try to combine them.

So that's the "why both".  As for which one to use?  Use `git_array_oid_t` internally.  Use `git_oidarray` if you need to return an array of OIDs to callers.

**Reply by punkymaniac** [reply]
And what about the usage of `git_oidarray` in input of function ? for example in `git_merge_base_many`:
```c
/* Current prototype: */
int git_merge_base_many(git_oid *out, git_repository *repo, size_t length, const git_oid [] input_array);

/* Why this function prototype is not: */
int git_merge_base_many(git_oid *out, git_repository *repo, const git_oidarray input_array);
```

I ask because `git_strarray` seem to be used in input and output of function but not `git_oidarray`.

**Reply by ethomson** [reply]
`git_oidarray` is an output type not an input type.  It's there so that we can allocate some memory and hand that back to callers and then give them a function to free it with.

It's not an input type because they don't need to a function to free an array of `git_oid`s that they've allocated themselves.  They can just use whatever `malloc` and `free` type functions they're already using.

The rationale for `git_strarray` should be similar.  If we're using it for input and not output, we should fix that!

**Reply by ethomson** [reply]
Oy, after looking at this a little bit more, changing `git_strarray` to a type that's only for returning data would be a bit challenging.  There are definitely bits of the API that are load bearing on a `git_strarray`.

I don't love that `git_strarray` _sometimes_ has our memory in it (and needs to be `git_strarray_free`d) and _sometimes_ has caller data in it (and should definitely _not_ be `git_strarray_free`d) but it still follows the principle: you allocate it?  You free it.

Moving from `git_strarray` to `const char **`, `size_t` doesn't have the same benefits that moving away from `git_buf` in our public API does.  `git_buf` is a core part of the internal API that was made a part of the external API.  That's the rationale behind big changes there, so that we can wall that up into public/private.  `git_strarray` doesn't really have the same problem.

So...  I'm going to defer fixing this.
